File: haskell/day1/Main.hs
import qualified Data.Set as Set
import Data.Maybe (fromJust)
import Data.List (scanl')


main :: IO ()
main = do
    frequency_changes <- parse <$> readFile "input.txt"
    putStrLn ("Part one: " ++ show (partOne frequency_changes))
    putStrLn ("Part two: " ++ show (partTwo frequency_changes))


parse :: String -> [Int]
parse = map parseInt . lines
  where
    parseInt ('+':xs) = read xs
    parseInt xs       = read xs


partOne :: [Int] -> Int
partOne = sum


partTwo :: [Int] -> Int
partTwo frequency_changes = head (dupes frequencies)
  where
    frequencies = scanl' (+) 0 (cycle frequency_changes)


dupes :: Ord a => [a] -> [a]
dupes = go Set.empty
  where
    go _    []                = []
    go seen (x:xs)
        | x `Set.member` seen = x : go (x `Set.insert` seen) xs
        | otherwise           = go (x `Set.insert` seen) xs

File: rust/day1/src/main.rs
use std::io::BufRead;
use std::io::BufReader;
use std::fs::File;
use std::collections::HashSet;


fn part_one(frequency_changes: &Vec<i32>) -> i32 {
    frequency_changes.iter().sum()
}


fn part_two(frequency_changes: &Vec<i32>) -> i32 {
    let mut current_freq: i32 = 0;
    let mut seen: HashSet<i32> = HashSet::new();
    for freq in frequency_changes.iter().cycle() {
        current_freq += freq;
        if seen.contains(&current_freq) {
            break;
        }
        seen.insert(current_freq);
    }
    current_freq
}


fn main() {
    let input_file = File::open("input.txt").unwrap();

    let frequency_changes: Vec<i32> =
        BufReader::new(&input_file)
        .lines()
        .map(|line| line.unwrap().parse().unwrap())
        .collect();

    println!("Part one: {:?}", part_one(&frequency_changes));
    println!("Part two: {:?}", part_two(&frequency_changes));
}
